core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -816,6 +816,10 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
     if (reassignment.isEmpty) {
       info("No more partitions need to be reassigned. Deleting zk path %s".format(ReassignPartitionsZNode.path))
       zkClient.deletePartitionReassignment()
+      // If the reassignment znode was set before we registered the watcher, we need to queue the PartitionReassignment
+      // event manually
+      if (zkClient.registerZNodeChangeHandlerAndCheckExistence(partitionReassignmentHandler))
+        eventManager.put(PartitionReassignment)
```

Comment:
Could we just do eventManager.put(PartitionReassignment)? PartitionReassignment.process() registers the existence watcher as the first thing. So, the logic can be handled there. i.e. (1) if the reassignment path is empty, an exist watcher will be left on the path; (2) if the reassignment path is not empty, it will be processed.

Reply:
@junrao's suggestion is simpler and works too.

One downside to his approach:
The `ControllerState` metric will have an extra blip of `PartitionReassignment` at some point after rebalance completes since there could be events between now and the enqueued `PartitionReassignment`.

Reply:
I did as Jun suggested, but I noticed one thing: we made things more efficient in this PR by avoiding one ZK read and watch registration, but we still have one remaining in `removePartitionFromReassignedPartitions`:

```
// read the current list of reassigned partitions from zookeeper
    val partitionsBeingReassigned = zkClient.getPartitionReassignment.mapValues(replicas => ReassignedPartitionsContext(replicas))
```

This means that we could actually have implemented this by keeping the `handleDeletion` implementation. It seemed a bit more obscure and brittle though so I didn't do that.

Reply:
Is there a reason why we don't update the znode based on the cached data?

Reply:
That's a good point. It seems that we can just remove the partition from the cached data, instead of reading the same value from ZK again. Do you want to include that in the patch?